src/i2s.rs

```diff
@@ -0,0 +1,1835 @@
+    }
+
+    impl Default for SlotBitWidth {
+        #[inline(always)]
+        fn default() -> Self {
+            Self::Auto
+        }
+    }
+
+    impl SlotBitWidth {
+        /// Convert this to the underlying SDK type.
+        #[inline(always)]
+        pub(crate) fn as_sdk(&self) -> i2s_slot_bit_width_t {
+            match self {
+                Self::Auto => 0,
+                Self::Bits8 => 8,
+                Self::Bits16 => 16,
+                Self::Bits24 => 24,
+                Self::Bits32 => 32,
+            }
+        }
+    }
+
+    impl TryFrom<u32> for SlotBitWidth {
+        type Error = EspError;
+
+        fn try_from(value: u32) -> Result<Self, Self::Error> {
+            match value {
+                0 => Ok(Self::Auto),
+                8 => Ok(Self::Bits8),
+                16 => Ok(Self::Bits16),
+                24 => Ok(Self::Bits24),
+                32 => Ok(Self::Bits32),
+                _ => Err(EspError::from(ESP_ERR_INVALID_ARG).unwrap()),
+            }
+        }
+    }
+
+    /// I2S channel slot mode.
+    #[derive(Clone, Copy, Eq, PartialEq)]
+    pub enum SlotMode {
+        /// Mono mode:
+        /// * When transmitting, transmit the same data in all slots.
+        /// * When receiving, only receive data from the first slot.
+        Mono,
+
+        /// Stereo mode:
+        /// * When transmitting, transmit different data in each slot.
+        /// * When receiving, receive data from all slots.
+        Stereo,
+    }
+
+    impl Default for SlotMode {
+        #[inline(always)]
+        fn default() -> Self {
+            Self::Stereo
+        }
+    }
+
+    impl SlotMode {
+        /// Convert this to the underlying SDK type.
+        #[inline(always)]
+        pub(crate) fn as_sdk(&self) -> i2s_slot_mode_t {
+            match self {
+                Self::Mono => 0,
+                Self::Stereo => 1,
+            }
+        }
+    }
+
+    pub trait MaybePin {
+        fn pin(&self) -> i32;
+    }
+
+    impl MaybePin for () {
+        fn pin(&self) -> i32 {
+            -1
+        }
+    }
+
+    impl<T: Pin> MaybePin for PeripheralRef<'static, T> {
+        fn pin(&self) -> i32 {
+            <T as Pin>::pin(self)
+        }
+    }
+}
+
+pub trait I2s: Send {
+    fn port() -> i2s_port_t;
+}
+
+/// The inner details about an I2S driver. This is allocated separately to prevent moving the atomics around.
+struct I2sDriverInternal {
+    i2s: u8,
+    config: config::Config,
+    rx_chan_handle: i2s_chan_handle_t,
+    tx_chan_handle: i2s_chan_handle_t,
+    rx_chan_available: AtomicBool,
+    tx_chan_available: AtomicBool,
+}
+
+/// Allow references for the driver to be send across threads.
+///
+/// This is necessary becuase i2s_chan_handle_t is a pointer, which is not Sync. However, the ESP-IDF API is thread
+/// safe here.
+///
+/// Note that we cannot implement Send. We do *not* want this struct to be sent between threads.
+unsafe impl Sync for I2sDriverInternal {}
```

Comment:
Why can't we implement Send? That should be pretty much almost always OK? The one case I recall this is not OK is if you have thread affinity code, which you don't.

Reply:
This should never be owned by a thread. That would be weird; the thread could then drop it. This should just last for `'static`.

Reply:
No longer an issue. The I2S peripheral can be owned; I2sDriver (now I2sStdDriver) can be owned and moved just like anything else.